README
fdsend README

Introduction
------------
fdsend is yet another file descriptor passing abstraction, specifically for
Python.  This package offers a few conveniences not commonly found together in
other abstractions:  sending multiple files at once, sending arbitrary data, and
working with both files and file descriptors.

License
-------
fdsend is free software, licensed under the GPLv2+.  It was originally developed
by Michael J. Pomraning (http://pilcrow.madison.wi.us/) and is currently
maintained by Philipp Kern and Fabian Knittel.  Small portions of fdsend are
adapted from other sources, see the main source file for details.

Requirements
------------
Python >= 2.2 is required, as is SCM_RIGHTS support on AF_UNIX sockets. Only
Linux has been tested.  Setuptools is required for automatic unit test runs at
build-time.

Installation
------------
  $ tar zxf fdsend-$VERSION.tar.gz
  $ cd fdsend-$VERSION
  $ python setup.py build
  $ python setup.py install

Details
-------
fdsend allows the passing of open files between unrelated processes via
local sockets (using SCM_RIGHTS), a process known as file descriptor
passing.  The following functions are available:

  sendfds()
  recvfds()
  socketpair()

Unlike some other simplifications of the sendmsg()/recvmsg() interface,
fdsend allows multiple files to be transferred in a single operation, and
permits ordinary socket messages to accompany the files.  Additionally,
fdsend understands bona fide Python sockets and files, as well as objects
implementing fileno() methods and integers representing file descriptors.

Errors are raised via the socket.error exception object.

Example
-------

  import fdsend
  import socket  # for socket.error

  # -- sender; send 12 files
  #
  nulldevs = [ file('/dev/null') for i in xrange(12) ]
  fdsend.sendfds(sock, "here you go!", fds = nulldevs)

  # -- recipient; receive up to 128-byte message, and up to 32 files
  #
  (msg, fds) = fdsend.recvfds(sock, 128, numfds = 32)
  import os
  fds = [ os.fdopen(fileno) for fileno in fds ]

  # -- exception handling
  #
  import socket
  try:
      notasock = file("/dev/null")
      fdsend.sendfds(notasock, "This won't work!")
  except socket.error, e:
      import errno
      if e[0] == errno.ENOTSOCK:
          print "What did you expect? :)"

Known Bugs
----------
- No provision is made to support msg_accrights systems (nor I_SENDFD fd
  passing).

Feedback
--------
Please report problems, bugs, feature requests, successes on the mailing
list <python-fdsend@googlegroups.com>.

Misc
----
- The usual limitation in "rights" send/recv fd implementations, wherein
  msg_iov conveys a single, often hardcoded, byte, is a venerable approach
  dating back at least as far as Stevens' APUE examples (1992).  For many
  applications, it's no limitation at all.
- socketpair() is only for Python versions lacking this under the 'socket'
  module.

Related Interest
----------------
- scgi: (python) contains a passfd module.
  (http://www.mems-exchange.org/software/scgi/)
- sendmsg: The sendmsg module (python) is a mostly complete sendmsg interface.
  (http://www.python.org/pycon/dc2004/papers/51/migration-code/sendmsg/)
- Socket::MsgHdr: (perl) offers a full sendmsg interface.
  (http://search.cpan.org/dist/Socket-MsgHdr/lib/Socket/MsgHdr.pm)
- bglibs:  provides a variety of UNIX-ish conveniences, including a
  socket_sendfd function.
  (http://untroubled.org/bglibs/)
- I_SENDFD ioctl: The I_SENDFD ioctl is an alternative approach for
  streams-based systems
  (http://pubs.opengroup.org/onlinepubs/007908799/xsh/ioctl.html)


NEWS
0.3.1 (21 Jan 2023)
  * Support Python >= 3.10.

0.3.0 (06 Jan 2020)
  * Support Python 3.

0.2.1 (23 Aug 2011)
  * Fixes integer conversion bug in recvfds and sendfds, found by
    Philipp Kern (via the Debian buildds).

0.2.0 (12 Aug 2011)
  * Added a test case that tests sendfds and recvfds.
  * Maintainership transferred to Philipp Kern and Fabian Knittel.

0.1 (03 Nov 2004)
  * First release by Michael J. Pomraning


test.py
#!/usr/bin/env python
# vim:set et ts=4 sw=4:
# GPLv2; (C) 2011 Philipp Kern <pkern@debian.org>
# Simple test case to verify that fdsend is working correctly.
# (I.e. both message and fds are sent properly.)  This fails
# on Debian Lenny/amd64, presumably due to a kernel bug.

# Lenny has Python 2.5, so we need this.
from __future__ import with_statement

import fdsend
import os
import socket
import sys
import tempfile
import time

def try_connect(s, fn, n=0):
    try:
        s.connect(fn)
    except socket.error:
        if n > 3:
            raise
        time.sleep(1)
        try_connect(s, fn, n+1)

def child(fn):
    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    try_connect(s, fn)
    req, fds = fdsend.recvfds(s, 1024, numfds=1)
    files = [os.fdopen(fd, 'r') for fd in fds]
    result = files[0].read()
    if req == "fds" and result == "TESTDATA\n":
        sys.exit(0)
    else:
        print "Test failed!"
        print "Message expected: fds"
        print "Message got: " + req
        print "FD content expected: TESTDATA"
        print "FD got: " + result.strip()
        sys.exit(1)

def parent(child_pid, fn):
    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    s.bind(fn)
    s.listen(1)
    (conn, address) = s.accept()

    with tempfile.TemporaryFile(mode='w+') as f:
        f.write("TESTDATA\n")
        f.flush()
        f.seek(0)
        fdsend.sendfds(conn, "fds", fds=[f])

    (pid, exit_status) = os.waitpid(child_pid, 0)
    if exit_status != 0:
        # Test failed
        sys.exit(1)

def main():
    temp_dir = tempfile.mkdtemp()
    socket_filename = os.path.join(temp_dir, 'sock')

    pid = os.fork()
    if pid == 0:
        child(socket_filename)
    else:
        parent(pid, socket_filename)

if __name__ == '__main__':
    main()



fdsend/tests.py
#!/usr/bin/env python
# vim:set et ts=4 sw=4:
"""
Provides test cases for :mod:`fdsend`.
"""
# Copyright (C) 2011 Philipp Kern <pkern@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import fdsend
import os
import socket
import tempfile
import unittest
import time
import shutil
import sys

def try_connect(sock, sock_fn, retries=30):
    """Try connecting with `sock` to `sock_fn`.  The connection is attempted
    `retries` times.
    """
    try_num = 1
    while True:
        try:
            sock.connect(sock_fn)
            return
        except socket.error:
            if try_num > retries:
                raise
            time.sleep(0.1)
            try_num += 1


class TestFDSend(unittest.TestCase):
    """Simple test case to verify that fdsend is working correctly. (I.e. both
    message and fds are sent properly.)  This fails on Debian Lenny/amd64,
    presumably due to a kernel bug.
    """
    # We cannot change the function names dictated by :class:`TestCase`.
    # pylint: disable=C0103

    longMessage = True
    DIRECT_DATA = "This data was sent directly via the socket."
    FILE_DATA = "This data was stored in a file whose file descriptor "\
            "was passed via the socket."

    def setUp(self):
        """Prepares a temporary directory and the name of the communication
        socket file.
        """
        self.temp_dir = tempfile.mkdtemp()
        self.sock_fn = os.path.join(self.temp_dir, 'sock')

    def tearDown(self):
        """Clears out the temporary directory.
        """
        shutil.rmtree(self.temp_dir, ignore_errors=True)

    def test_send_receive(self):
        """Test sending and receiving of direct data and file descriptors.

        It splits off a child process to test the fdsend module while
        conveniently using blocking calls.
        """
        pid = os.fork()
        if pid == 0:
            # The child creates the socket and sends data.
            try:
                self.send_data()
            except: # pylint: disable=W0702
                # Suppress all exception handling to avoid :class:`TestCase`
                # from kicking in.
                os._exit(1) # pylint: disable=W0212
            os._exit(0) # pylint: disable=W0212

        # The parent receives the data and verifies it.
        self.receive_data()

        # Wait for the child to exit.
        (pid, exit_status) = os.waitpid(pid, 0)
        self.assertTrue(exit_status == 0)

    def receive_data(self):
        """This method verifies that the data received through the UNIX domain
        socket is as expected.
        """
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        # As the other process might not have created the socket file yet, try
        # connecting repeatedly.
        try_connect(sock, self.sock_fn)

        data, fds = fdsend.recvfds(sock, 1024, numfds=64)
        sock.close()

        # Now verify that what we received is what we expected.

        # Exactly one file descriptor.
        self.assertEqual(len(fds), 1)

        # Received correct direct data.
        if sys.version_info >= (3,):
            wanted_data = bytes(self.DIRECT_DATA, encoding='ascii')
        else:
            wanted_data = self.DIRECT_DATA
        self.assertEqual(data, wanted_data)

        # Verify file descriptor position.
        transferred_fp = os.fdopen(fds[0], 'r')
        self.assertEqual(transferred_fp.tell(), len(self.FILE_DATA))

        # Verify data in passed file descriptor.
        transferred_fp.seek(0)
        self.assertEqual(transferred_fp.read(), self.FILE_DATA)
        transferred_fp.close()

    def send_data(self):
        """Creates a new UNIX socket named `sock_fn` and a temporary file and
        transmits some data and the file descriptor of the temporary file
        through that socket.
        """
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.bind(self.sock_fn)
        sock.listen(1)
        client_sock, _ = sock.accept()
        sock.close()

        tmp_fp = open(os.path.join(self.temp_dir, 'somefile'), 'w+')
        tmp_fp.write(self.FILE_DATA)
        tmp_fp.flush()
        fdsend.sendfds(client_sock, self.DIRECT_DATA, fds=[tmp_fp])
        tmp_fp.close()
        # tmp_fp's file will be cleaned up by deleting the temporary directory.

        client_sock.close()

    def test_negative_recv(self):
        """fdsend.recvfds does not accept negative buffer lengths.
        """
        self.assertRaises(ValueError, fdsend.recvfds, 0, -1)


_fdsend.c
/* fdsend - SCM_RIGHTS file descriptor passing for Python.
 *
 * Copyright (C) 2004 Michael J. Pomraning <mjp{AT}pilcrow.madison.wi.us>
 * Copyright (C) 2011 Philipp Kern <pkern@debian.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Portions of this file -- preliminary defines and socketpair() routine --
 * are derived from scgi version 1.2, which is is Copyright (c) 2004
 * Corporation for National Research Initiatives; All Rights Reserved.
 */

#define PY_SSIZE_T_CLEAN
#include "Python.h"

struct module_state {
	PyObject *socketmodule_error;
};

#if PY_MAJOR_VERSION >= 3
#define IS_PY3K
#define INITERROR return NULL
#define GETSTATE(m) ((struct module_state*)PyModule_GetState(m))
#else
#define INITERROR return
#define GETSTATE(m) (&_state)
static struct module_state _state;
#endif

#ifndef __OpenBSD__
#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 500
#endif
#ifndef _XOPEN_SOURCE_EXTENDED
#define _XOPEN_SOURCE_EXTENDED 1 /* Solaris <= 2.7 needs this too */
#endif
#endif /* __OpenBSD__ */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include <stddef.h>

/* obj2fd
 *
 * A PyArg_Parse... format converter.
 */
static int
obj2fd(PyObject *o, void *p) {
	int fd = -1;

	if (o == Py_None)
		goto okay;

	if ((fd = PyObject_AsFileDescriptor(o)) == -1)
		return 0;
okay:
	*(int *)p = fd;

	return 1;
}

/* pack_control
 *
 * Cram a sequence of open files (or file descriptors) into the ancillary
 * data of a msghdr.  Note that the msg_control member is dynamically
 * allocated by this function, and may be freed by a call to
 * free_packed_control().
 */
static int
pack_control(PyObject *seq, struct msghdr *msg) {
	PyObject *fast_seq = NULL;
	int i, sz;
	int *fd_ptr = NULL;
	struct cmsghdr *cmsg;
	int ok = 0;

	fast_seq = PySequence_Fast(seq, "files argument must be a sequence");
	if (NULL == fast_seq) return 0;

	sz = PySequence_Fast_GET_SIZE(fast_seq);
	if (0 == sz) {
		msg->msg_controllen = 0;
		msg->msg_control = NULL;
		return 1;
	}
	msg->msg_controllen = CMSG_SPACE(sizeof(int) * sz);
	msg->msg_control = PyMem_Malloc(msg->msg_controllen);
	if (NULL == msg->msg_control) return 0;

	cmsg = CMSG_FIRSTHDR(msg);
	cmsg->cmsg_len = CMSG_LEN(sizeof(int) * sz);
	cmsg->cmsg_level = SOL_SOCKET;
	cmsg->cmsg_type = SCM_RIGHTS;
	fd_ptr = (int *)CMSG_DATA(cmsg);
	for (i = 0; i < sz; i++) {
		PyObject *f;
		int fd;

		f = PySequence_Fast_GET_ITEM(fast_seq, i);
		if (f == NULL) goto error;
		if ((fd = PyObject_AsFileDescriptor(f)) == -1) goto error;
		*fd_ptr = fd;
		fd_ptr++;
	}
	ok = 1;

out:
	Py_DECREF(fast_seq);
	return ok;

error:
	ok = 0;
	if (msg->msg_control) PyMem_Free(msg->msg_control);
	goto out;
}

static void
free_packed_control(struct msghdr *msg)
{
	if (NULL == msg) return;
	if (msg->msg_control) PyMem_Free(msg->msg_control);
	return;
}

/* unpack_control
 *
 * Unpack the ancillary data, assuming SCM_RIGHTS (file descriptor passing).
 *
 * Be prepared for:
 *   - more than one cmsghdr
 *   - more than one fd per cmsghdr
 */
static PyObject *
unpack_control(struct msghdr *msg) {
	struct cmsghdr *cmsg;
	int i = 0;
	PyObject *tup = NULL;

	for (cmsg = CMSG_FIRSTHDR(msg);
	     cmsg != NULL;
	     cmsg = CMSG_NXTHDR(msg,cmsg)) {
		if (cmsg->cmsg_level != SOL_SOCKET
		    || cmsg->cmsg_type != SCM_RIGHTS) {
			PyErr_SetString(PyExc_RuntimeError,
					"Unexpected cmsg level or type found");
			goto error;
		}
		i += (cmsg->cmsg_len - sizeof(struct cmsghdr))/sizeof(int);
	}

	tup = PyTuple_New(i);
	if (NULL == tup) goto error;

	i = 0;
	for (cmsg = CMSG_FIRSTHDR(msg);
	     cmsg != NULL; 
	     cmsg = CMSG_NXTHDR(msg,cmsg)) {
		int j;
		int *pfd = (int *)CMSG_DATA(cmsg);
		int upper_bound;
		upper_bound = cmsg->cmsg_len
			      - sizeof(struct cmsghdr) 
			      - sizeof(int);
		for (j = 0; j <= upper_bound; j += sizeof(int)) {
			PyTuple_SET_ITEM(tup, i, PyLong_FromLong((long)pfd[i]));
			i++;
		}
	}
	return tup;
error:
	if (tup) { Py_DECREF(tup); }

	return NULL;
}

static PyObject *
fdsend_sendfds(PyObject *m, PyObject *args, PyObject *kw)
{
	struct msghdr mh;
	struct iovec iov;
	int fd, r, flags = 0;
	Py_ssize_t msg_len;
	PyObject *fds = Py_None;

	static char *keywords[] = {"fd", "msg", "flags", "fds", 0};
	if (!PyArg_ParseTupleAndKeywords(args, kw, "O&s#|iO:sendfds",
					 keywords,
					 obj2fd, &fd,
					 &iov.iov_base, &msg_len,
					 &flags,
					 &fds))
		return NULL;
	iov.iov_len = msg_len;

	memset(&mh, '\0', sizeof(mh));
	mh.msg_iov = &iov;
	mh.msg_iovlen = 1;

	if (fds != Py_None) {
		if (! pack_control(fds, &mh)) return NULL;
	}
	Py_BEGIN_ALLOW_THREADS
	r = sendmsg(fd, &mh, flags);
	Py_END_ALLOW_THREADS
	free_packed_control(&mh);

	if (r < 0)
		return PyErr_SetFromErrno(GETSTATE(m)->socketmodule_error);

#ifndef IS_PY3K
	return PyInt_FromLong((long) r);
#else
	return PyLong_FromLong((long) r);
#endif
}

static PyObject *
fdsend_recvfds(PyObject *m, PyObject *args, PyObject *kw)
{
	struct msghdr mh;
	struct iovec iov;
	struct cmsghdr *cmsg;
	PyObject *buf;
	int numfds = 32;
	int fd, flags = 0;
	size_t r;
	int msg_len;
	PyObject *ret = NULL;

	static char *keywords[] = {"fd", "len", "flags", "numfds", 0};
	if (!PyArg_ParseTupleAndKeywords(args, kw, "O&i|ii:recvfds", keywords,
					obj2fd, &fd,
					&msg_len,
					&flags,
					&numfds))
		return NULL;
	if (msg_len < 0) {
		PyErr_SetString(PyExc_ValueError, "Negative len passed to recvfds");
		return NULL;
	}
	iov.iov_len = msg_len;

	memset(&mh, '\0', sizeof(mh));

	if (numfds > 0) {
		mh.msg_controllen = CMSG_SPACE(sizeof(int) * numfds);
		mh.msg_control = PyMem_Malloc(mh.msg_controllen);
		if (NULL == mh.msg_control) return NULL;
	}

#ifndef IS_PY3K
	buf = PyString_FromStringAndSize((char *) 0, iov.iov_len);
#else
	buf = PyBytes_FromStringAndSize((char *) 0, iov.iov_len);
#endif
	if (NULL == buf) goto error;
#ifndef IS_PY3K
	iov.iov_base = (void *)PyString_AS_STRING(buf);
#else
	iov.iov_base = (void *)PyBytes_AS_STRING(buf);
#endif
	/* uncomment the following for clearer strace(1)ing */
	/* memset(iov.iov_base, '\0', iov.iov_len); */

	mh.msg_iov = &iov;
	mh.msg_iovlen = 1;

	Py_BEGIN_ALLOW_THREADS
	r = recvmsg(fd, &mh, flags);
	Py_END_ALLOW_THREADS

	if (r < 0) {
		PyErr_SetFromErrno(GETSTATE(m)->socketmodule_error);
		goto error;
	}

	if (r != iov.iov_len)
#ifndef IS_PY3K
		_PyString_Resize(&buf, r);
#else
		_PyBytes_Resize(&buf, r);
#endif
	cmsg = CMSG_FIRSTHDR(&mh);
	if (NULL == cmsg
	    || cmsg->cmsg_level != SOL_SOCKET
	    || cmsg->cmsg_type != SCM_RIGHTS)
		return Py_BuildValue("(N())", buf);

	ret = Py_BuildValue("(OO)", buf, unpack_control(&mh));

out:
	if (mh.msg_control) PyMem_Free(mh.msg_control);
	return ret;

error:
	if (buf) { Py_DECREF(buf); }
	goto out;
}

static char fdsend_sendfds__doc__[] =
"sendfds(fd, msg, flags=0, fds=None) -> bytes_sent\n"
"\n"
"Send msg across the socket represented by fd, optionally accompanied by a\n"
"sequence (tuple or list) of open file handles.  For example:\n"
"\n"
"  >>> devnull = file(\"/dev/null\")\n"
"  >>> sendfds(the_socket, \"null device\", (devnull,))\n"
"\n"
"The socket fd and members of the fds sequence may be any representation\n"
"described in the module docstring.\n"
"\n"
"Note that most underlying implementations require at least a one byte msg\n"
"to transmit open files.";

static char fdsend_recvfds__doc__[] =
"recvfds(fd, len, flags=0, numfds=64) -> (message, fd_tuple)\n"
"\n"
"Receive a message of up to length len and up to numfds new files from socket\n"
"object fd.\n"
"\n"
"Though the socket object may be given as any of the representations listed\n"
"in the module docstring, new files returned in fd_tuple are always integral\n"
"file descriptors.  See os.fdopen for a means of transforming them into\n"
"Python file objects.\n"
"\n"
"There is presently no way to detect msg_flags values (e.g., MSG_CTRUNC).";

static char socketpair__doc__[] =
"socketpair(family, type, proto=0) -> (fd, fd)\n"
"\n"
"Provided as a convenience for Python versions lacking a socket.socketpair\n"
"implementation.";

static PyObject *
fdsend_socketpair(PyObject *self, PyObject *args)
{
	int family, type, proto=0;
	int fd[2];

	if (!PyArg_ParseTuple(args, "ii|i:socketpair", &family, &type, &proto))
		return NULL;

	if (socketpair(family, type, proto, fd) < 0) {
		PyErr_SetFromErrno(PyExc_IOError);
		return NULL;
	}

	return Py_BuildValue("(ii)", (long) fd[0], (long) fd[1]);
}


/* List of functions */

static PyMethodDef fdsend_methods[] = {
	{"sendfds", (PyCFunction)fdsend_sendfds,
		METH_VARARGS|METH_KEYWORDS, fdsend_sendfds__doc__},
	{"recvfds", (PyCFunction)fdsend_recvfds,
		METH_VARARGS|METH_KEYWORDS, fdsend_recvfds__doc__},
	{"socketpair",	fdsend_socketpair, METH_VARARGS, socketpair__doc__},
	{NULL,		NULL}		/* sentinel */
};

static char module__doc__[] =
"fdsend allows the passing of open files between unrelated processes via\n"
"local sockets (using SCM_RIGHTS), a process known as file descriptor\n"
"passing.  The following functions are available:\n"
"\n"
"  sendfds()\n"
"  recvfds()\n"
"  socketpair()\n"
"\n"
"Unlike some other simplifications of the sendmsg()/recvmsg() interface,\n"
"fdsend allows multiple files to be transferred in a single operation, and\n"
"permits ordinary socket messages to accompany the files.  Additionally,\n"
"fdsend understands bona fide Python sockets and files, as well as objects\n"
"implementing fileno() methods and integers representing file descriptors.\n"
"\n"
"Errors are raised via the socket.error exception object.";

#ifdef IS_PY3K
static int _fdsend_traverse(PyObject *m, visitproc visit, void *arg) {
	Py_VISIT(GETSTATE(m)->socketmodule_error);
	return 0;
}

static int _fdsend_clear(PyObject *m) {
	Py_CLEAR(GETSTATE(m)->socketmodule_error);
	return 0;
}

static struct PyModuleDef moduledef = {
	PyModuleDef_HEAD_INIT,
	"_fdsend",
	module__doc__,
	sizeof(struct module_state),
	fdsend_methods,
	NULL,
	_fdsend_traverse,
	_fdsend_clear,
	NULL
};
#endif

#ifndef IS_PY3K
DL_EXPORT(void)
init_fdsend(void)
#else
PyObject *
PyInit__fdsend(void)
#endif
{
	PyObject *m, *sm;

	/* Create the module and add the functions and documentation */
#ifndef IS_PY3K
	m = Py_InitModule3("_fdsend", fdsend_methods, module__doc__);
#else
	m = PyModule_Create(&moduledef);
#endif

	if (m == NULL) {
		INITERROR;
	}
	/* Fetch reference to socket.error class. */
	if ((sm = PyImport_ImportModule("socket")) != NULL) {
		GETSTATE(m)->socketmodule_error =
			PyObject_GetAttrString(sm, "error");
	}

#ifdef IS_PY3K
	return m;
#endif
}
